core/src/main/scala/org/apache/spark/scheduler/TaskAssigner.scala

```diff
@@ -0,0 +1,233 @@
+  def construct(workOffer: Seq[WorkerOffer]): Unit = {
+    offer = workOffer.map(o => new OfferState(o))
+  }
+
+  // Invoked at each round of Taskset assignment to initialize the internal structure.
+  def init(): Unit
+
+  // Whether there is offer available to be used inside of one round of Taskset assignment.
+  def hasNext: Boolean
+
+  // Returned the next assigned offer based on the task assignment strategy.
+  def getNext(): OfferState
+
+  // Invoked by the TaskScheduler to indicate whether the current offer is accepted or not so that
+  // the assigner can decide whether the current worker is valid for the next offering.
+  def offerAccepted(assigned: Boolean): Unit
+
+  // Invoked at the end of resource offering to release internally maintained resources.
+  // Subclass is responsible to release its own private resources.
+  def reset(): Unit = {
+    offer = null
+  }
+}
+
+object TaskAssigner extends Logging {
+  private val roundrobin = classOf[RoundRobinAssigner].getCanonicalName
+  private val packed = classOf[PackedAssigner].getCanonicalName
+  private val balanced = classOf[BalancedAssigner].getCanonicalName
+  private val assignerMap: Map[String, String] =
+    Map("roundrobin" -> roundrobin,
+      "packed" -> packed,
+      "balanced" -> balanced)
+
+  def init(conf: SparkConf): TaskAssigner = {
+    val assignerName = conf.get(config.SPARK_SCHEDULER_TASK_ASSIGNER.key, "roundrobin")
+      .toLowerCase()
+    val className = assignerMap.getOrElse(assignerName, roundrobin)
+    val CPUS_PER_TASK = conf.getInt("spark.task.cpus", 1)
+    val assigner = try {
+      logInfo(s"Constructing an assigner as $className")
+      Utils.classForName(className).getConstructor()
+        .newInstance().asInstanceOf[TaskAssigner]
+    } catch {
+      case _: Throwable =>
+        logInfo(s"$assignerName cannot be constructed, fallback to default $roundrobin.")
+        new RoundRobinAssigner()
+    }
+    assigner.withCpuPerTask(CPUS_PER_TASK)
+    assigner
+  }
+}
+
+/**
+ * Assign the task to workers with available cores in roundrobin manner.
+ */
+class RoundRobinAssigner extends TaskAssigner {
+  private var idx = 0
+
+  override def construct(workOffer: Seq[WorkerOffer]): Unit = {
+    offer = Random.shuffle(workOffer.map(o => new OfferState(o)))
+  }
+
+  override def init(): Unit = {
+    idx = 0
+  }
+
+  override def hasNext: Boolean = idx < offer.size
+
+  override def getNext(): OfferState = {
+    offer(idx)
+  }
+
+  override def offerAccepted(assigned: Boolean): Unit = {
+    idx += 1
+  }
+
+  override def reset(): Unit = {
+    super.reset
+    idx = 0
+  }
+}
+
+/**
+ * Assign the task to workers with the most available cores.
+ */
+class BalancedAssigner extends TaskAssigner {
+  private var maxHeap: PriorityQueue[OfferState] = _
+  private var currentOffer: OfferState = _
+
+  override def construct(workOffer: Seq[WorkerOffer]): Unit = {
+    offer = Random.shuffle(workOffer.map(o => new OfferState(o)))
+  }
```

Comment:
As you will put offers into the `PriorityQueue`, is it still necessary to do shuffling?